File: marketing/channels/reddit.js
// =============================================================================
// Knol Marketing — Reddit Channel Adapter
// Posts to subreddits via Reddit API (free tier: 60 req/min, 100 posts/day)
// =============================================================================

const https = require('https');
const querystring = require('querystring');

const CONFIG = {
  API_BASE: 'https://oauth.reddit.com',
  AUTH_URL: 'https://www.reddit.com/api/v1/access_token',
  RATE_LIMIT: { perDay: 10, perMinute: 60 },
  TARGET_SUBREDDITS: [
    { name: 'rust', type: 'technical', minKarma: 10 },
    { name: 'programming', type: 'general', minKarma: 50 },
    { name: 'MachineLearning', type: 'ml', minKarma: 50 },
    { name: 'opensource', type: 'oss', minKarma: 10 },
    { name: 'selfhosted', type: 'infra', minKarma: 10 },
    { name: 'artificial', type: 'ai', minKarma: 10 },
    { name: 'LocalLLaMA', type: 'llm', minKarma: 10 },
  ],
};

function getJson(path, token) {
  return new Promise((resolve) => {
    const req = https.request({
      hostname: 'oauth.reddit.com',
      path,
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
        'User-Agent': 'knol-marketing/0.1.0',
      },
    }, (res) => {
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => {
        try {
          resolve({
            success: true,
            json: JSON.parse(data),
            statusCode: res.statusCode,
            headers: res.headers,
          });
        } catch (e) {
          resolve({ success: false, error: `Parse error: ${e.message}`, statusCode: res.statusCode });
        }
      });
    });
    req.on('error', (e) => resolve({ success: false, error: e.message }));
    req.end();
  });
}

// Get OAuth2 access token (script app type)
async function authenticate(credentials) {
  const { clientId, clientSecret, username, password } = credentials;

  if (!clientId || !clientSecret || !username || !password) {
    return { success: false, error: 'Missing Reddit credentials' };
  }

  const body = querystring.stringify({
    grant_type: 'password',
    username,
    password,
  });

  return new Promise((resolve) => {
    const auth = Buffer.from(`${clientId}:${clientSecret}`).toString('base64');
    const req = https.request({
      hostname: 'www.reddit.com',
      path: '/api/v1/access_token',
      method: 'POST',
      headers: {
        'Authorization': `Basic ${auth}`,
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent': 'knol-marketing/0.1.0',
        'Content-Length': Buffer.byteLength(body),
      },
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          if (parsed.access_token) {
            resolve({
              success: true,
              token: parsed.access_token,
              expiresIn: parsed.expires_in,
              statusCode: res.statusCode,
              headers: res.headers,
            });
          } else {
            resolve({
              success: false,
              error: parsed.error || 'Auth failed',
              rateLimited: res.statusCode === 429,
              statusCode: res.statusCode,
              headers: res.headers,
            });
          }
        } catch (e) {
          resolve({ success: false, error: `Parse error: ${e.message}` });
        }
      });
    });

    req.on('error', (e) => resolve({ success: false, error: e.message }));
    req.write(body);
    req.end();
  });
}

// Submit a link or self post
async function submitPost(subreddit, title, content, token, kind = 'self') {
  const body = querystring.stringify({
    sr: subreddit,
    kind,
    title,
    ...(kind === 'self' ? { text: content } : { url: content }),
    resubmit: true,
    send_replies: true,
  });

  return new Promise((resolve) => {
    const req = https.request({
      hostname: 'oauth.reddit.com',
      path: '/api/submit',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent': 'knol-marketing/0.1.0',
        'Content-Length': Buffer.byteLength(body),
      },
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          if (parsed.success || (parsed.json?.data?.url)) {
            resolve({
              success: true,
              url: parsed.json?.data?.url || parsed.url,
              postId: parsed.json?.data?.id,
              statusCode: res.statusCode,
              headers: res.headers,
            });
          } else {
            const errors = parsed.json?.errors || [];
            resolve({
              success: false,
              error: errors.length ? errors.map(e => e.join(': ')).join('; ') : `HTTP ${res.statusCode}`,
              rateLimited: res.statusCode === 429,
              statusCode: res.statusCode,
              headers: res.headers,
            });
          }
        } catch (e) {
          resolve({ success: false, error: `Parse error: ${e.message}` });
        }
      });
    });

    req.on('error', (e) => resolve({ success: false, error: e.message }));
    req.write(body);
    req.end();
  });
}

// Post a comment on an existing thread
async function postComment(thingId, text, token) {
  const body = querystring.stringify({
    thing_id: thingId,
    text,
  });

  return new Promise((resolve) => {
    const req = https.request({
      hostname: 'oauth.reddit.com',
      path: '/api/comment',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent': 'knol-marketing/0.1.0',
        'Content-Length': Buffer.byteLength(body),
      },
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        resolve({
          success: res.statusCode === 200,
          statusCode: res.statusCode,
        });
      });
    });

    req.on('error', (e) => resolve({ success: false, error: e.message }));
    req.write(body);
    req.end();
  });
}

async function findEngagementOpportunities(token, options = {}) {
  const query = options.query || '(memory OR rag OR vector OR "ai agents" OR llm)';
  const subreddits = Array.isArray(options.subreddits) && options.subreddits.length
    ? options.subreddits
    : ['rust', 'MachineLearning', 'LocalLLaMA', 'opensource', 'selfhosted'];
  const perSubLimit = Math.max(1, Math.min(10, options.limitPerSub || 4));
  const minScore = Number.isFinite(options.minScore) ? options.minScore : 3;
  const minComments = Number.isFinite(options.minComments) ? options.minComments : 0;

  const all = [];

  for (const subreddit of subreddits) {
    const path = `/r/${encodeURIComponent(subreddit)}/search.json?q=${encodeURIComponent(query)}&restrict_sr=1&sort=new&t=week&limit=${perSubLimit}`;
    const response = await getJson(path, token);
    if (!response.success || !response.json?.data?.children) continue;

    for (const item of response.json.data.children) {
      const d = item?.data || {};
      if ((d.score || 0) < minScore) continue;
      if ((d.num_comments || 0) < minComments) continue;
      if (d.archived || d.locked) continue;
      all.push({
        id: d.id,
        name: d.name, // Fullname (e.g. t3_abc123), required for comments.
        title: d.title,
        subreddit: d.subreddit,
        score: d.score || 0,
        comments: d.num_comments || 0,
        author: d.author || '',
        createdUtc: d.created_utc || null,
        url: d.url || '',
        permalink: d.permalink ? `https://reddit.com${d.permalink}` : '',
      });
    }
  }

  all.sort((a, b) => (b.score + b.comments) - (a.score + a.comments));
  return {
    success: true,
    opportunities: all.slice(0, Math.max(1, Math.min(20, options.limit || 10))),
  };
}

module.exports = { authenticate, submitPost, postComment, findEngagementOpportunities, CONFIG };

File: knol-enterprise/crates/service-marketing/src/channels/reddit.rs
//! Reddit channel adapter — OAuth2 password grant → POST /api/submit.

use tracing::{info, warn};

use super::{PublishContent, PublishResult};
use crate::config::ChannelCredentials;
use crate::error::MarketingError;

/// Authenticate with Reddit using password grant.
async fn authenticate(
    http_client: &reqwest::Client,
    credentials: &ChannelCredentials,
) -> Result<String, MarketingError> {
    let client_id =
        credentials
            .reddit_client_id
            .as_ref()
            .ok_or_else(|| MarketingError::Channel {
                channel: "reddit".into(),
                message: "Reddit credentials not configured".into(),
            })?;
    let client_secret = credentials.reddit_client_secret.as_deref().unwrap_or("");
    let username = credentials.reddit_username.as_deref().unwrap_or("");
    let password = credentials.reddit_password.as_deref().unwrap_or("");

    let resp = http_client
        .post("https://www.reddit.com/api/v1/access_token")
        .basic_auth(client_id, Some(client_secret))
        .header("User-Agent", "knol-marketing/0.1.0")
        .form(&[
            ("grant_type", "password"),
            ("username", username),
            ("password", password),
        ])
        .send()
        .await
        .map_err(|e| MarketingError::Channel {
            channel: "reddit".into(),
            message: format!("Auth failed: {}", e),
        })?;

    let body: serde_json::Value = resp.json().await.map_err(|e| MarketingError::Channel {
        channel: "reddit".into(),
        message: format!("Auth parse error: {}", e),
    })?;

    body["access_token"]
        .as_str()
        .map(|s| s.to_string())
        .ok_or_else(|| MarketingError::Channel {
            channel: "reddit".into(),
            message: format!("No access_token in response: {}", body),
        })
}

pub async fn publish(
    content: &PublishContent,
    http_client: &reqwest::Client,
    credentials: &ChannelCredentials,
) -> Result<PublishResult, MarketingError> {
    let token = authenticate(http_client, credentials).await?;

    let subreddit = content.subreddit.as_deref().unwrap_or("rust");
    let title = content.title.as_deref().unwrap_or(&content.text);

    let params = vec![
        ("sr", subreddit.to_string()),
        ("kind", "self".to_string()),
        ("title", title.to_string()),
        (
            "text",
            content.body.as_deref().unwrap_or(&content.text).to_string(),
        ),
        ("resubmit", "true".to_string()),
    ];

    let resp = http_client
        .post("https://oauth.reddit.com/api/submit")
        .header("Authorization", format!("Bearer {}", token))
        .header("User-Agent", "knol-marketing/0.1.0")
        .form(&params)
        .send()
        .await
        .map_err(|e| MarketingError::Channel {
            channel: "reddit".into(),
            message: e.to_string(),
        })?;

    let status = resp.status();
    let body: serde_json::Value = resp.json().await.unwrap_or_default();

    let post_url = body["json"]["data"]["url"].as_str().map(|s| s.to_string());
    let post_id = body["json"]["data"]["id"].as_str().map(|s| s.to_string());

    if body["success"].as_bool().unwrap_or(false) || post_url.is_some() {
        info!("Reddit: posted to r/{}", subreddit);
        Ok(PublishResult::success("reddit", post_id, post_url))
    } else if status.as_u16() == 429 {
        warn!("Reddit: rate limited");
        Err(MarketingError::RateLimited {
            channel: "reddit".into(),
            current: 0,
            limit: 0,
            window: "api".into(),
        })
    } else {
        let errors = &body["json"]["errors"];
        let err_msg = if errors.is_array() && !errors.as_array().unwrap().is_empty() {
            format!("{}", errors)
        } else {
            format!("HTTP {}", status)
        };
        warn!("Reddit: {}", err_msg);
        Ok(PublishResult::failure("reddit", err_msg))
    }
}
